SPEC.md

```diff
@@ -47,7 +47,15 @@ The key of the map is a filename and is a CBOR text string encoded in UTF-8.
 The value of the map is another CBOR map with the following standard fields:
 
   - `type`: String with value of `'dir'`.
-  - `data`: TODO: Define structure for directory content data.
+  - `data`: Object or Link. 
+    - `data` as Object: values must be links to `file` or `dir` nodes.
+    - `data` as Link: Link must resolve to a `hamt` root node.
   - `size`: Integer. Cumulative size of all directories and files in `data`.
 
 The `type` field set to `dir`.
+
+The `data` field can be either an Object or a Link. When representing a directory structure with an object
+the keys are the directory names and the values must be links to either `file` or `dir` nodes. When 
+the keyspace is too large for a single node `data` should be a link to `hamt` node.
+
+`hamt` has not yet been specified but will follow the implementation [here](https://github.com/ipfs/go-hamt-ipld).
```

Comment:
Should we make this mandatory? It is kind of cumbersome that users have to enable/disable `hamt` today (most of them miss it or don't understand its value).

Reply:
For small directories it would be a big performance hit to require `hamt` in every directory definition.

One thing I was thinking about was simply removing from this spec entirely the notion of what is in a single block vs. multiple blocks. If we used paths (instead of "properties") across the spec and stated clearly that any path could be a link, an implementation could potentially collapse the whole directory structure into a single block. Then we'd carve out an exception for `hamt` to be treated as a flat namespace. This would all be much easier to require if the path lookups for `hamt` could be implemented in wasm and attached to the root node but that's still a long way out.

Thoughts? @Stebalien @vmx

Reply:
Using paths instead of properties sounds like it aligns well with what IPLD is about.

Reply:
I'll work on a PR for a path based spec after I land this :)

Reply:
My thoughts from brewing over in schemaland is I'd like to have something specified like:

```
union UnixfsNode = UnixfsFile | UnixfsDir | #....

struct UnixfsDir = {
	name optional Fname
	entries {Flabel:UnixfsNode}<CHAMT>
	size Integer
}
```

(The part to highlight here is of course `UnixfsDir.entries`; the rest is just for some contextualization.)

- `entries` is the name of the field in this structure
- the `{` indicates we're about to describe a map
- `Flabel` is the name of the map key type ("label" vs "name" to account for the (unrelated, here) issues about path name character ranges...)
- `UnixfsNode` is the type of values in the map
- and `<CHAMT>` is an *Advanced Layout* specifier.

I did some more drafting of what "Advanced Layout" could mean in a general sense in this gist -- https://gist.github.com/warpfork/6df17e791936d1f9b0d5e5483678c8bf#file-gistfile1-txt-L42-L57 -- while exploring other related schema thoughts.

In writing that, it seemed to me that we could very easily say that implementations of an Advanced Layout can easily do an "inline"/simple mode for small data.

Does this seem valid?  Does it give us a way to punt this issue to something less specific to UnixFS spec'ing?